src/txHashBuilder.c

```diff
@@ -1047,11 +1077,93 @@ static void txHashBuilder_assertCanLeaveValidityIntervalStart(tx_hash_builder_t*
 	}
 }
 
-void txHashBuilder_finalize(tx_hash_builder_t* builder, uint8_t* outBuffer, size_t outSize)
+void txHashBuilder_enterMint(tx_hash_builder_t* builder)
 {
 	_TRACE("state = %d", builder->state);
 
 	txHashBuilder_assertCanLeaveValidityIntervalStart(builder);
+	{
+		// Enter mint
+		BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, TX_BODY_KEY_MINT);
+	}
+	builder->state = TX_HASH_BUILDER_IN_MINT;
+}
+
+void txHashBuilder_addMint_topLevelData(
+        tx_hash_builder_t* builder, uint16_t numAssetGroups
+)
+{
+	_TRACE("state = %u", builder->state);
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_MINT);
+
+	builder->multiassetData.remainingAssetGroups = numAssetGroups;
+	// Map(numAssetGroups)[
+	//   { * policy_id => { * asset_name => uint } }
+	// ]
+	BUILDER_APPEND_CBOR(CBOR_TYPE_MAP, numAssetGroups);
+	if (0 != numAssetGroups) {
+		builder->state = TX_HASH_BUILDER_IN_MINT_ASSET_GROUP;
+	} else {
+		// TODO is this an error case?
+	}
+}
+
+void txHashBuilder_addMint_tokenGroup(
+        tx_hash_builder_t* builder,
+        const uint8_t* policyIdBuffer, size_t policyIdSize,
+        uint16_t numTokens
+)
+{
+	addTokenGroup(builder, policyIdBuffer, policyIdSize, numTokens,
+	              TX_HASH_BUILDER_IN_MINT_ASSET_GROUP, TX_HASH_BUILDER_IN_MINT_TOKEN);
+}
+
+void txHashBuilder_addMint_token(
+        tx_hash_builder_t* builder,
+        const uint8_t* assetNameBuffer, size_t assetNameSize,
+        int64_t amount
+)
+{
+	//  TODO int64 -> uint64 conversion should work fine ?
```

Comment:
Only if the int is non-negative?

Reply:
I think this describes it pretty well. The signed is simply interpreted as an unsigned, and that's just fine for us.
https://stackoverflow.com/questions/50605/signed-to-unsigned-conversion-in-c-is-it-always-safe

Reply:
I don't understand. A negative int64 must not be serialized as `CBOR_TYPE_UNSIGNED`, it is `CBOR_TYPE_NEGATIVE`. This is how we are supposed to serialize integers:
https://datatracker.ietf.org/doc/html/rfc7049#section-3.9
If they are too big (I'm not sure now), we might need something like
https://datatracker.ietf.org/doc/html/rfc7049#section-2.4.2

Reply:
I see what you mean, right now negative values always take up the maximum 8 bytes, since as unsigned they are always larger than VALUE_MIN_W8. I'm not clear on who will read this CBOR (and the hash made from it) and who we want to be consistent with. I'll implement the negative side of CBOR.

Reply:
It would be great if we had examples of transactions with negative mint values, including ones that do not fit in uint32. Cardano-cli might actually not be complying with the CBOR spec, as we have discovered for token ordering. Please consult this with @gabrielKerekes who probably already has such txs.

Reply:
LedgerJS already tests all the way to INT64_MIN, but I don't remember if my manual test with the generated hash value of minting was about a negative amount or not. But I think the more interesting case is where it _does_ fit into 32 bit, since that's where a difference should come up.